src/context/products_context.js

```diff
@@ -13,18 +13,72 @@ import {
   GET_SINGLE_PRODUCT_ERROR,
 } from '../actions'
 
-const initialState = {}
+const initialState = {
+  isSidebarOpen: false,
+  products_loading: false,
+  products_error: false,
+  products: [],
+  featured_products: [],
+  single_product_loading: false,
+  single_product_error: false,
+  single_product: {},
+}
 
 const ProductsContext = React.createContext()
 
 export const ProductsProvider = ({ children }) => {
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  const openSidebar = () => {
+    dispatch({ type: SIDEBAR_OPEN })
+  }
+  const closeSidebar = () => {
+    dispatch({ type: SIDEBAR_CLOSE })
+  }
+
+  const fetchProducts = async (url) => {
+    dispatch({ type: GET_PRODUCTS_BEGIN })
+    try {
+      const response = await axios.get(url)
+      const products = response.data.products
+      dispatch({ type: GET_PRODUCTS_SUCCESS, payload: products })
+    } catch (error) {
+      dispatch({ type: GET_PRODUCTS_ERROR })
+    }
+  }
+//fn that fetches single product- without line 55 I used to get an array of products just like in fetchProducts fn , now it gives an error in console bundle.js:146779     GET http://localhost:3000/products/NaN 404 (Not Found)- so in SingleProductPage.js i converted it to number - line 18
+  const fetchSingleProduct = async (url,id) => {
+    dispatch({ type: GET_SINGLE_PRODUCT_BEGIN })
+    try {
+      const response = await axios.get(url/id)
+      const products = response.data.products
```

Comment:
The return value from the API is an object with the root field \`product\` SINGULAR. For example: 

[https://ecommerce-6kwa.onrender.com/api/v1/products/64804bb82f146aaa74a85c3a](https://ecommerce-6kwa.onrender.com/api/v1/products/64804bb82f146aaa74a85c3a)

returns:

```json
{
  "product": {
    "_id": "64804bb82f146aaa74a85c3a",
    "name": "Kitchen",
    "price": 50000,
    "description": "Cloud bread VHS hell of banjo bicycle rights jianbing umami mumblecore etsy 8-bit pok pok +1 wolf. Vexillologist yr dreamcatcher waistcoat, authentic chillwave trust fund. Viral typewriter fingerstache pinterest pork belly narwhal. Schlitz venmo everyday carry kitsch pitchfork chillwave iPhone taiyaki trust fund hashtag kinfolk microdosing gochujang live-edge",
    "image": "/uploads/example.jpg",
    "category": "bedroom",
    "company": "ikea",
    "colors": [
      "#0000ff",
      "#000"
    ],
    "featured": false,
    "freeShipping": false,
    "inventory": 15,
    "averageRating": 3,
    "numOfReviews": 2,
    "user": "647dcbf9ffdc226f332b7842",
    "createdAt": "2023-06-07T09:19:52.091Z",
    "updatedAt": "2023-06-07T16:22:33.047Z",
    "__v": 0,
    "reviews": [
      {
        "_id": "64804c7a2f146aaa74a85c46",
        "rating": 5,
        "title": "WOW product",
        "comment": "Very very bad product",
        "user": "647dcbf9ffdc226f332b7842",
        "product": "64804bb82f146aaa74a85c3a",
        "createdAt": "2023-06-07T09:23:06.207Z",
        "updatedAt": "2023-06-07T14:03:01.855Z",
        "__v": 0
      },
      {
        "_id": "64807f4897e9abfa727f045a",
        "rating": 1,
        "title": "Great quality",
        "comment": "Very very bad product",
        "user": "647dcc71ffdc226f332b784e",
        "product": "64804bb82f146aaa74a85c3a",
        "createdAt": "2023-06-07T12:59:52.826Z",
        "updatedAt": "2023-06-07T12:59:52.826Z",
        "__v": 0
      }
    ],
    "id": "64804bb82f146aaa74a85c3a"
  }
}
```

so to access that we'd want to do: `response.data.product` instead of `response.data.products`

Reply:
@akosasante THANK YOU